src/components/include/utils/shared_ptr.h

```diff
@@ -36,353 +36,18 @@
-    const SharedPtr<OtherObjectType>& Other) {
-  dropReference();
-
-  mObject = Other.mObject;
-  mReferenceCounter = Other.mReferenceCounter;
-
-  if (0 != mReferenceCounter) {
-    atomic_post_inc(mReferenceCounter);
-  }
-
-  return *this;
-}
-
-template <typename ObjectType>
-template <typename OtherObjectType>
-utils::SharedPtr<OtherObjectType> utils::SharedPtr<
-    ObjectType>::static_pointer_cast(const SharedPtr<ObjectType>& pointer) {
-  SharedPtr<OtherObjectType> casted_pointer;
-  casted_pointer.mObject = static_cast<OtherObjectType*>(pointer.mObject);
-  casted_pointer.mReferenceCounter = pointer.mReferenceCounter;
-
-  if (0 != casted_pointer.mReferenceCounter) {
-    atomic_post_inc(casted_pointer.mReferenceCounter);
-  }
-
-  return casted_pointer;
-}
-
-template <typename ObjectType>
-template <typename OtherObjectType>
-utils::SharedPtr<OtherObjectType> utils::SharedPtr<
-    ObjectType>::dynamic_pointer_cast(const SharedPtr<ObjectType>& pointer) {
-  SharedPtr<OtherObjectType> casted_pointer;
-  casted_pointer.mObject = dynamic_cast<OtherObjectType*>(pointer.mObject);
-  if (NULL != casted_pointer.mObject) {
-    casted_pointer.mReferenceCounter = pointer.mReferenceCounter;
-
-    if (0 != casted_pointer.mReferenceCounter) {
-      atomic_post_inc(casted_pointer.mReferenceCounter);
-    }
-  }
-
-  return casted_pointer;
-}
-
-template <typename ObjectType>
-ObjectType* utils::SharedPtr<ObjectType>::operator->() const {
-  DCHECK(mObject);
-  return mObject;
-}
-
-template <typename ObjectType>
-ObjectType& utils::SharedPtr<ObjectType>::operator*() const {
-  DCHECK(mObject);
-  return *mObject;
-}
-
-template <typename ObjectType>
-utils::SharedPtr<ObjectType>::operator bool() const {
-  return valid();
-}
-
-template <typename ObjectType>
-void utils::SharedPtr<ObjectType>::reset() {
-  reset_impl(0);
-}
-
-template <typename ObjectType>
-void utils::SharedPtr<ObjectType>::reset(ObjectType* other) {
-  DCHECK(other != NULL);
-  reset_impl(other);
-}
-
-template <typename ObjectType>
-void SharedPtr<ObjectType>::release() {
-  deleter_(mObject);
-  mObject = 0;
-
-  delete mReferenceCounter;
-  mReferenceCounter = 0;
-}
-
-template <typename ObjectType>
-void utils::SharedPtr<ObjectType>::reset_impl(ObjectType* other) {
-  dropReference();
-  mObject = other;
-  mReferenceCounter = new uint32_t(1);
-}
-
-template <typename ObjectType>
-inline void SharedPtr<ObjectType>::dropReference() {
-  if (0 != mReferenceCounter) {
-    if (1 == atomic_post_dec(mReferenceCounter)) {
-      release();
-    }
-  }
-}
-
-template <typename ObjectType>
-ObjectType* SharedPtr<ObjectType>::get() const {
-  return mObject;
-}
-
-template <typename ObjectType>
-inline bool SharedPtr<ObjectType>::valid() const {
-  if (mReferenceCounter && (0 < *mReferenceCounter)) {
-    return (mObject != NULL);
-  }
-  return false;
+/* Test if the given shared ptr is valid and in use
+ */
+template <typename T>
```

Comment:
Might do well with a rename of this file, with this being the only function in the file